fw/cache.c

```diff
@@ -328,21 +328,46 @@ tfw_cache_key_node(unsigned long key)
 	return key % num_online_nodes();
 }
 
+/**
+ * Release memory dynamically allocated for every node cpus
+ */
+static void
+tfw_release_node_cpus(void)
+{
+	int node;
+
+	for_each_online_node(node) {
+		if(c_nodes[node].nr_cpus > 0)
+			kfree(c_nodes[node].cpu);
```

Comment:
Is this a protection against garbage in `c_nodes[node].cpu`? Otherwise, it's safe to `kfree()` any pointer, even a null pointer, without an extra check.

Also, after freeing it's highly desirable that you keep both `.nr_cpu` and `.cpu` consistent with the new state of being unused (i.e. zero). Again, if this state is maintained properly, then you don't need that extra check above.

An aside question: does this (and other) code handle CPUs that can be put online or offline, or CPU migration, or when the node that was online is now offline? Also, perhaps, you should `kfree()` whatever you have allocated and not just rely on these `_online_node()`, etc, functions?

Reply:
1. Yes, but this code will be removed in the next version. 
2. Yes again, but  `c_nodes` also will be dynamically allocated, so we don't need it. But need to check `c_nodes`.
3. Currently assumed no CPU manipulation, just decreasing memory consumption to use it only for online nodes/CPUs.

Reply:
Thanks. At first glance, I didn't find any stipulated requirements regarding CPU manipulation in Tempesta Wiki. If there is, perhaps, it should be made more obvious?

Reply:
> Thanks. At first glance, I didn't find any stipulated requirements regarding CPU manipulation in Tempesta Wiki. If there is, perhaps, it should be made more obvious?

I don't know. I definitely saw such information, but don't remember where is.